File: benchmarks/rust/json_parse.rs
// JSON parse benchmark — parse array of objects, sum "value" fields
// Manual parser (no serde — single-file benchmark, no cargo)
use std::fmt::Write;

const NUM_OBJECTS: usize = 50000;

fn generate_json(n: usize) -> String {
    let mut s = String::with_capacity(n * 40);
    s.push('[');
    for i in 0..n {
        if i > 0 { s.push(','); }
        write!(s, r#"{{"id":{},"value":{}}}"#, i, 100 + (i % 1000)).unwrap();
    }
    s.push(']');
    s
}

fn parse_and_sum(json: &[u8]) -> i64 {
    let mut total: i64 = 0;
    let len = json.len();
    let mut i = 0;
    // Scan for "value": pattern
    while i + 8 < len {
        if json[i] == b'"' && json[i+1] == b'v' && json[i+2] == b'a'
            && json[i+3] == b'l' && json[i+4] == b'u' && json[i+5] == b'e'
            && json[i+6] == b'"' && json[i+7] == b':' {
            i += 8;
            let mut val: i64 = 0;
            let mut neg = false;
            if i < len && json[i] == b'-' { neg = true; i += 1; }
            while i < len && json[i] >= b'0' && json[i] <= b'9' {
                val = val * 10 + (json[i] - b'0') as i64;
                i += 1;
            }
            if neg { val = -val; }
            total += val;
        } else {
            i += 1;
        }
    }
    total
}

fn main() {
    let json = generate_json(NUM_OBJECTS);
    let sum = parse_and_sum(json.as_bytes());
    println!("len={} sum={}", json.len(), sum);
}

File: benchmarks/c/json_parse.c
/* JSON parse benchmark — parse array of objects, sum "value" fields
 * Input: self-generated JSON string ~1MB
 * Format: [{"id":0,"value":100},{"id":1,"value":101},...]
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#define NUM_OBJECTS 50000

/* Generate JSON string in-memory */
static char *generate_json(int n, size_t *out_len) {
    /* Estimate: each object ~30 bytes, plus brackets/commas */
    size_t cap = (size_t)n * 40 + 16;
    char *buf = malloc(cap);
    size_t pos = 0;
    buf[pos++] = '[';
    for (int i = 0; i < n; i++) {
        if (i > 0) buf[pos++] = ',';
        pos += sprintf(buf + pos, "{\"id\":%d,\"value\":%d}", i, 100 + (i % 1000));
    }
    buf[pos++] = ']';
    buf[pos] = '\0';
    *out_len = pos;
    return buf;
}

/* Minimal JSON parser — extract "value" fields and sum them */
static int64_t parse_and_sum(const char *json, size_t len) {
    int64_t total = 0;
    const char *p = json;
    const char *end = json + len;
    /* Scan for "value": pattern */
    while (p < end - 8) {
        if (p[0] == '"' && p[1] == 'v' && p[2] == 'a' && p[3] == 'l'
            && p[4] == 'u' && p[5] == 'e' && p[6] == '"' && p[7] == ':') {
            p += 8;
            /* Parse integer */
            int64_t val = 0;
            int neg = 0;
            if (*p == '-') { neg = 1; p++; }
            while (p < end && *p >= '0' && *p <= '9') {
                val = val * 10 + (*p - '0');
                p++;
            }
            if (neg) val = -val;
            total += val;
        } else {
            p++;
        }
    }
    return total;
}

int main(void) {
    size_t len;
    char *json = generate_json(NUM_OBJECTS, &len);
    int64_t sum = parse_and_sum(json, len);
    printf("len=%zu sum=%ld\n", len, sum);
    free(json);
    return 0;
}
